File: packages/media_processing_kit/android/src/main/kotlin/com/bittercamera/media_processing_kit/EffectProcessor.kt
package com.bittercamera.media_processing_kit

/**
 * 效果处理器接口（由 B/C 实现）：
 * - D 的媒体管线会把 task 分发给指定 effectId 对应的 processor
 * - processor 自己决定是否需要解码/逐帧处理/编码（MVP 可先简单实现）
 *
 * 注意：
 * - 这里先只定义“最小可用”的 file-in/file-out 接口，便于尽快联调。
 * - 未来要做真正逐帧处理，可以扩展为 surface/texture/bytebuffer 级别。
 */
interface EffectProcessor {
    fun processVideo(taskId: String, inputPath: String, outputPath: String, config: Map<String, Any?>, callbacks: TaskCallbacks)
    fun processImage(taskId: String, inputPath: String, outputPath: String, config: Map<String, Any?>, callbacks: TaskCallbacks)
    fun cancel(taskId: String)
}

interface TaskCallbacks {
    fun onProgress(taskId: String, progress: Double)
    fun onCompleted(taskId: String, outputPath: String)
    fun onError(taskId: String, errorCode: String, errorMessage: String)
}

File: packages/media_processing_kit/ios/Classes/EffectProcessor.swift
import Foundation

/// 效果处理器协议（由 B/C 实现）：
/// - D 只负责任务管理、事件回调、文件 I/O（MVP 默认 passthrough）
/// - 真正的苦瓜脸效果由 B/C 通过 register(effectId:processor:) 注入
public protocol EffectProcessor {
  func processVideo(taskId: String, inputPath: String, outputPath: String, config: [String: Any?], callbacks: TaskCallbacks)
  func processImage(taskId: String, inputPath: String, outputPath: String, config: [String: Any?], callbacks: TaskCallbacks)
  func cancel(taskId: String)
}

public protocol TaskCallbacks {
  func onProgress(taskId: String, progress: Double)
  func onCompleted(taskId: String, outputPath: String)
  func onError(taskId: String, errorCode: String, errorMessage: String)
}

public final class EffectRegistry {
  public static let shared = EffectRegistry()
  private var processors: [String: EffectProcessor] = [:]
  private let lock = NSLock()

  public func register(effectId: String, processor: EffectProcessor) {
    lock.lock(); defer { lock.unlock() }
    processors[effectId] = processor
  }

  public func get(effectId: String) -> EffectProcessor? {
    lock.lock(); defer { lock.unlock() }
    return processors[effectId]
  }
}

